src/algorithms/calorimetry/CalorimeterHitDigi.cc

```diff
@@ -166,41 +157,34 @@ void CalorimeterHitDigi::single_hits_digi(){
 
         const long long tdc = std::llround((time + m_normDist(generator) * tRes) * stepTDC);
 
-        if (eDep> 1.e-3) m_log->trace("E sim {} \t adc: {} \t time: {}\t maxtime: {} \t tdc: {} \t cell ID {}", eDep, adc, time, m_capTime, tdc, ahit->getCellID());
-        auto rawhit = new edm4hep::RawCalorimeterHit(
-                ahit->getCellID(),
+        if (eDep> 1.e-3) m_log->trace("E sim {} \t adc: {} \t time: {}\t maxtime: {} \t tdc: {} \t cell ID {}", eDep, adc, time, m_capTime, tdc, ahit.getCellID());
+        rawhits->create(
+                ahit.getCellID(),
                 (adc > m_capADC ? m_capADC : adc),
                 tdc
         );
-        rawhits.push_back(rawhit);
     }
+
+    return std::move(rawhits);
 }
 
-//------------------------
-// signal_sum_digi
-//------------------------
-void CalorimeterHitDigi::signal_sum_digi( void ){
+std::unique_ptr<edm4hep::RawCalorimeterHitCollection> CalorimeterHitDigi::signal_sum_digi(const edm4hep::SimCalorimeterHitCollection &simhits)  {
+    std::unique_ptr<edm4hep::RawCalorimeterHitCollection> rawhits { std::make_unique<edm4hep::RawCalorimeterHitCollection>() };
 
     // find the hits that belong to the same group (for merging)
     std::unordered_map<long long, std::vector<const edm4hep::SimCalorimeterHit*>> merge_map;
-    for (auto ahit : simhits) {
-        int64_t hid = ahit->getCellID() & id_mask;
+    for (const auto &ahit : simhits) {
+        int64_t hid = ahit.getCellID() & id_mask;
 
-        m_log->trace("org cell ID in {:s}: {:#064b}", m_readout, ahit->getCellID());
+        m_log->trace("org cell ID in {:s}: {:#064b}", m_readout, ahit.getCellID());
         m_log->trace("new cell ID in {:s}: {:#064b}", m_readout, hid);
 
-        auto    it  = merge_map.find(hid);
-
-        if (it == merge_map.end()) {
-            merge_map[hid] = {ahit};
-        } else {
-            it->second.push_back(ahit);
-        }
+        merge_map[hid].push_back(&ahit);
```

Comment:
Careful storing a raw pointer in `merge_map`. I tried something similar to this in another algorithm, but eventually found this following pattern (storing the index) to be safer:
```cpp
for(std::size_t i=0; i<simhits.size(); i++) {
  auto ahit = simhits.at(i);
  // ...
  merge_map[hid].push_back(i);
}
```
I'm afraid I don't remember the details, but storing raw pointers in some map like this was causing me to have sporadic segfaults and similar problems.

Reply:
Yes, you are pointing to a bug. Collections don't allow you to get an actual pointer. I don't know if anything other than using an index can be done here.